include/libdnf/advisory/advisory_query.hpp

```diff
@@ -78,11 +76,11 @@ class AdvisoryQuery : public AdvisorySet {
     void filter_reference(
         const std::string & pattern,
         sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ,
-        const std::optional<std::string> type = {});
+        const std::string & type = {});
```

Comment:
I am not sure whether this is good approach or not. First of all - do we have a list of valid types or this is a string in metadata that we try to match in field type? From the description it looks like that as a type only three strings are valid.

What other option do we have:

1. Overloaded methods:

```
    void filter_reference(
        const std::string & pattern,
        sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ,
        const std::optional<std::string> type = {});
```
```
    void filter_reference(
        const std::string & pattern,
        sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ,
        const std::optional<std::string> type = {});
        const std::string & type);
```
2. Raw pointer (not nice)
```
    void filter_reference(
        const std::string & pattern,
        sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ,
        const std::optional<std::string> type = {});
        const std::string * type);
```
3. Splitting reference filters ()
```
void filter_reference_name()
void filter_reference_type()
void filter_reference_name_type() - to do both together is good for performance 
```
4. default value
```
   void filter_reference(
        const std::string & pattern,
        sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ,
        const std::optional<std::string> type = {});
        const std::string & type = {});
```
May I ask the community for the feedback?

Reply:
Definitely, I'd be glad to have it discuss in a broader scope. And also to make things consistently when doing any changes in future. Another way, maybe, since we have a several valid values could be defining an enum for this.

Reply:
Regarding enum - I'm really not sure that the valid values are defined anywhere. I did a bit of searching and didn't find any standard describing updateinfo (well, I found this opensuse standard - https://en.opensuse.org/openSUSE:Standards_Rpm_Metadata_UpdateInfo which defines only "bugzilla" and "cvs" values). I'm afraid that different distributions can use it's own set of types. And from libsolv POV the reference type looks like an arbitrary string.

In this case I would go with overloaded `filter_reference(pattern, cmp_type)` and `filter_reference(pattern, cmp_type, type)` methods.

Reply:
Originally the types were an ENUM but I have removed them in https://github.com/rpm-software-management/libdnf/pull/1365.

I agree with @m-blaha, since we don't have any proper definition of metadata, it can be any string just like type or severity of advisory itself.

Reply:
Thanks for the comments! I also like the overloads approach for this case.

Reply:
I am just curios, where in DNF4 we require reference type? Or what kind of user case we have for DNF5?

Reply:
I am asking because there could be multiple types provided, then there could be a vector with strings as an argument.

Reply:
@j-mracek Yes, there could be third overloaded version accepting the vector. But I consider it out of scope of this PR. The TODO comment will do here.

Reply:
I found two different usages of advisory references in our code
```
    if (with_bz->get_value()) {
        advisories.filter_reference("*", libdnf::sack::QueryCmp::IGLOB, {"bugzilla"});
    }
    if (with_cve->get_value()) {
        advisories.filter_reference("*", libdnf::sack::QueryCmp::IGLOB, {"cve"});
    }
```
From the first usage I recommend to add `filter_reference_type` (name may be different). According to usage I recommend vector<string> as an argument for `void filter_reference_type()` because the command requires relation between types `or` rather then `and` - If I am correct advisory has only one type of reference therefore and will always result in empty set (currently implemented with `and` operator).

For testing I would recommend to play with `dnf5 advisory list --with-bz --with-cve`

The second usage uses both types and reference name string and it is correct to have there only one type - string not a vector. What about to have type as a mandatory argument?

Reply:
@jan-kolarik I am fine with overloaded method. I am sorry for complication. I need to know how we use the method and during investigation I discovered strange behavior. I can create issue to not forget to change behavior of the advisory command. Is it ok?

Reply:
Yes, it sounds even better to me having the issue instead of the TODO. I just wasn't sure if it's not an overkill for this case :)